sphinx/builders/__init__.py

```diff
@@ -90,6 +95,9 @@ def __init__(self, app: "Sphinx") -> None:
         self.tags.add("format_%s" % self.format)
         self.tags.add("builder_%s" % self.name)
 
+        # cache of Docutils Publishers (file type -> publisher object)
+        self._publishers: Dict[str, docutils.core.Publisher] = {}
```

Comment:
I prefer to store (cached) objects during the build to the registry (`sphinx.registry.SphinxComponentRegistry`). It's used as the common component store. I believe it's the more appropriate place. Could you move this? Then we can  keep using `sphinx.io.read_doc()`.

Reply:
I looked at the component registry but everything there seems to be storing types or functions, not instantiated classes. 

If you'd still prefer to move it I will do (I could add a helper `create_publisher` to the registry too)

A

Reply:
My large concern is not to add a new responsibility to the application object. Because it has already had a lot of responsibilities (in other words, methods and attributes). I feel creating a parser object is not an important responsibility of the "application" itself.

At present, the "reading" responsibility and its implementations are gathers to "sphinx.io" module. I believe it should not be moved to the app. This is why the cache should be stored to the registry. It's not important to use registry itself. But it should not be implemented to the app object.

Reply:
I moved the set-up to the registry, how is it now?

A